### ISSUE TITLE
Out of memory

### ISSUE THREAD

**Issue by giulia-berto** [brainlife/app-pyafq-segment#7]
I got this error: "slurmstepd: error: Detected 1 oom-kill event(s) in StepId=1033826.batch. Some of your processes may have been killed by the cgroup out-of-memory handler." The requested memory was above 40GB, do you think we should increase that value?

**Comment by soichih** [comment]
Could you post the memory usage graph? Also, how many fibers are in the input data?

**Comment by giulia-berto** [comment]
![memory-graphs](https://user-images.githubusercontent.com/20837136/149169267-23257a05-5137-4b6e-8cfe-f1b8bad46f9f.png)

The number of streamlines is 675000, thus not so many.

**Comment by soichih** [comment]
This App has a pretty low success rate (14.4%) I am not sure what's the cause of failures, but I guess we could try increasing it to say 100G and how much does the job actually take, and benchmark how much the App actually uses and figure out if that's normal?

**Comment by giulia-berto** [comment]
Yes, I agree. Also, it doesn't run on any shared resources, so I wonder whether it is still under development @bacaron. If so, I can help finalize it.

**Comment by bacaron** [comment]
Hey @soichih and @giulia-berto yeah this app is still "under development" and the failure rate is mostly due to 1) me and 2) oom and/or timeout errors. It's been on my to-do list for a while to correct these and I was planning on getting back to it when I return in February. But if either, or both, of you would want to take lead on this that would be fantastic!

As of now, there's a few issues with this app that need fixing in order for it to actually be useful:

1. Loading the tractograms, especially large tractograms (> 1 million streamlines), is a very time-consuming and memory-consuming process. For example, loading a 3 million streamline tractogram takes about 30 minutes on my laptop (4 core; 8 thread; max CPU speed 3900MHz; 20GB RAM). This is also a known issue amongst the community. Here's a github issue discussing the issue, but with .trk instead of .tck: https://github.com/nipy/nibabel/issues/943. However, I've found that the issue remains the same in .tck and that the issue hasn't been fully patched as of December. @soichih and I have had many discussions about this issue in the past. Currently, this is the biggest "rate limiting factor" of the app currently, and I'm not sure there's a quick or easy way to solve it.

2. Currently, I believe the app is downloading the ROI templates from somewhere which can also eat up some walltime. Basically the solution just requires downloading the templates to the app repo and pointing to them in the script (I think). Either way, I think this is also a rate limiting factor of the app.

3. Finally, I've hard-coded the warping of the DWI data to MNI space required for performing the segmentation. Currently I do this with just a single reg.syn_register_dwi command, but the example guide online (https://yeatmanlab.github.io/pyAFQ/auto_examples/plot_tract_profile.html#register-the-individual-data-to-a-template) suggests doing an affine first. However, i just made the assumption that DWI was already 'ACPC-aligned', i.e. affine'd to the MNI template and did not include this step. This, along with the syn warp, could be optional steps performed when a boolean configuration parameter is selected. I believe the affine code is linked in that example doc so you could probably just swipe it from there and add if statements to the python code for both the affine and the non-linear warp.

As of now, I believe that's all the main issues holding this app back. All of them, except for maybe 1, is addressable and manageable given adequate time I believe.

If either of you, or both, want to tackle this now that would be awesome! However if neither have enough time I can get back to this when I come back from my break!

Thanks!

**Comment by giulia-berto** [comment]
Thanks @bacaron for the comprehensive reply, this is very useful! For issue 1, I may have a solution, which is to use the fast loader that Emanuele was suggesting in the github issue you pointed out. I'd like to help, but I think this is not a priority right now (I was the one who started this issue, but I was just testing the App, it's not urgent for me). So maybe for now we should just increase the memory as @soichih suggested, and keep in mind this discussion for the future.

**Comment by soichih** [comment]
I've increased vmem request to 60gb to see what happens.
